mempool/src/pool.rs

```diff
@@ -0,0 +1,330 @@
+use std::cmp::Ord;
+use std::cmp::Ordering;
+use std::collections::BTreeSet;
+use std::collections::HashMap;
+use std::collections::HashSet;
+use std::fmt::Debug;
+use std::rc::Rc;
+
+use parity_scale_codec::Encode;
+use thiserror::Error;
+
+use common::chain::transaction::Transaction;
+use common::chain::transaction::TxInput;
+use common::chain::OutPoint;
+use common::primitives::amount::Amount;
+use common::primitives::Id;
+use common::primitives::Idable;
+use common::primitives::H256;
+
+// TODO this willbe defined elsewhere (some of limits.rs file)
+const MAX_BLOCK_SIZE_BYTES: usize = 1_000_000;
+
+const MEMPOOL_MAX_TXS: usize = 1_000_000;
+
+pub trait Mempool<C> {
+    fn create(chain_state: C) -> Self;
+    fn add_transaction(&mut self, tx: Transaction) -> Result<(), MempoolError>;
+    fn get_all(&self) -> Vec<&Transaction>;
+    fn contains_transaction(&self, tx: &Id<Transaction>) -> bool;
+    fn drop_transaction(&mut self, tx: &Id<Transaction>);
+    fn new_tip_set(&mut self) -> Result<(), MempoolError>;
+    fn get_chain_state(&self) -> &C;
+}
+
+// We split the ChainState interface into two parts:
+// 1. The "read-only" part of the interface, ChainStateView, is accessible to both the mempool (to validate transactoins) and the tests (to create mock transactions).
+// 2. The "write" part, ChainStateModify, is exposed ONLY to the mempool, not the tests. This is to prevent tests from modifying the ChainState object directly
+pub trait ChainState: ChainStateView + ChainStateModify {}
```

Comment:
I'm not sure I understand the motivation behind having two different interfaces. Isn't rust mutability of &self supposed to solve this problem? If you cast an interface to a non-mutable reference, you basically can't change it, even at compile-time.

Reply:
You're right. I came to the conclusion they are not necessary, so now there is only one interface.